core/unix/rseq_linux.c

```diff
@@ -151,6 +164,59 @@ rseq_cs_free(dcontext_t *dcontext, void *data)
     } while (record != NULL);
 }
 
+void
+rseq_check_glibc_support()
+{
+    /* We have already determined glibc's rseq support. */
+    if (glibc_rseq_offset > 0)
+        return;
+    module_iterator_t *iter = module_iterator_start();
+    while (module_iterator_hasnext(iter)) {
+        module_area_t *ma = module_iterator_next(iter);
+        int *rseq_offset_addr =
+            (int *)dr_get_proc_address((module_handle_t)ma->start, "__rseq_offset");
```

Comment:
We could use drsym here, but using extension libraries in core DR isn't really ideal. Another option is to change how we detect glibc's rseq support. Maybe based on the location of the first rseq syscall?

Reply:
Per offline discussion the plan is to see whether we can make the current heuristic search over TLS work

Reply:
Since the glibc-registered struct rseq is in struct pthread (which is == thread pointer IIUC) (https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/descr.h;h=0fa0e48f39a82b44bec899bceeeafc89ef7506b5;hb=HEAD#l409), we could hardcode that offset. I don't think it is guaranteed that the existing pthread fields will never change, but should be likely so. We can also add a test in our CI that confirms that that offset is still valid based on __rseq_offset.

Reply:
If we really don't want to hardcode that offset, we could maybe search all aligned offsets in struct pthread in the same manner we today search the loader's static TLS.

Reply:
pthread layout changing is part of what breaks Android support across releases and is now a big concern with glibc 2.34's merging of pthread into libc.  Xref `_android_v6_pthread_internal_t` vs `_android_v5_pthread_internal_t` and probably changes with later versions too but nobody has put in that support.  If can have a runtime check that can verify the offset and search nearby slots if it's wrong that seems better than a single assumed offset.

Reply:
I decided to not hardcode the offset at all. Though let me know if you feel it'll be better to add it anyway.

Reply:
So for x86: we search [-188, 836]. Glibc's struct rseq is on the positive side.

For AArch64: we search [-262457, 583]. Glibc's struct rseq is on the negative side.

Hmm... especially for AArch64, looks like it's worth trying the expected offset first.

Reply:
That is a big range for aarch64.

Reply:
I tried on some VMs, and the ranges look very different.

On a newer AArch64 machine (with glibc 2.35), the range was [-63, 320], whereas on another x86 machine (with older glibc), it was [-262332, 579].

Reply:
Added the hardcoded offset, and then the fallback of searching the wider space. I considered searching static TLS first, but decided against added more complexity.